Mathlib/Data/Sym/Sym2.lean

```diff
@@ -386,6 +386,87 @@ theorem map_congr {f g : α → β} {s : Sym2 α} (h : ∀ x ∈ s, f x = g x) :
+satisfy `p`, using the proof to apply `f`.
+-/
+def pmap {P : α → Prop} (f : ∀ a, P a → β) (s : Sym2 α) : (∀ a ∈ s, P a) → Sym2 β :=
+  let g : (p : α × α) → (∀ a ∈ Sym2.mk p, P a) → Sym2 β :=
+    fun p H => s(f p.1 (H p.1 <| mem_mk_left _ _), f p.2 (H p.2 <| mem_mk_right _ _))
+  Quot.recOn s g fun p q hpq => funext fun Hq => by
+    rw [rel_iff'] at hpq
+    have Hp : ∀ a ∈ Sym2.mk p, P a := fun a hmem =>
+      Hq a ((Sym2.mk_eq_mk_iff.2 hpq) ▸ hmem : a ∈ Sym2.mk q)
+    have h : ∀ {s₂ e H}, Eq.ndrec (motive := fun s => (∀ a ∈ s, P a) → Sym2 β) (g p) (b := s₂) e H =
+      g p Hp := by
+      rintro s₂ rfl _
+      rfl
+    refine h.trans (Quot.sound ?_)
+    rw [rel_iff', Prod.mk.injEq, Prod.swap_prod_mk]
+    apply hpq.imp <;> rintro rfl <;> simp
+
+@[simp]
+theorem forall_mem_pair {P : α → Prop} {a b : α} : (∀ x ∈ s(a, b), P x) ↔ P a ∧ P b := by
+  simp only [mem_iff, forall_eq_or_imp, forall_eq]
+
+lemma pmap_pair {P : α → Prop} (f : ∀ a, P a → β) (a b : α) (h : P a) (h' : P b) :
+    s(f a h, f b h') = pmap f s(a, b) (forall_mem_pair.mpr ⟨h, h'⟩) := rfl
+
+lemma pmap_pair' {P : α → Prop} (f : ∀ a, P a → β) (a b : α) (h : ∀ x ∈ s(a, b), P x) :
+    pmap f s(a, b) h = s(f a (h a (mem_mk_left a b)), f b (h b (mem_mk_right a b))) := rfl
+
+@[simp]
+lemma mem_pmap_iff {P : α → Prop} (f : ∀ a, P a → β) (z : Sym2 α) (h : ∀ a ∈ z, P a) (b : β) :
+    b ∈ z.pmap f h ↔ ∃ (a : α) (ha : a ∈ z), b = f a (h a ha) := by
+  induction' z with x y
+  rw [pmap_pair' f x y h]
+  simp only [mem_iff]
+  constructor
+  · rintro (rfl | rfl)
+    · use x, Or.inl rfl
+    · use y, Or.inr rfl
+  · rintro ⟨a, (rfl | rfl), rfl⟩
+    exact Or.inl rfl
+    exact Or.inr rfl
+
+lemma pmap_eq_map {P : α → Prop} (f : α → β) (z : Sym2 α) (h : ∀ a ∈ z, P a) :
+    z.pmap (fun a _ => f a) h = z.map f := by
+  induction' z with x y
+  rfl
+
+lemma map_pmap {Q : β → Prop} (f : α → β) (g : ∀ b, Q b → γ) (z : Sym2 α) (h' : ∀ b ∈ z.map f, Q b):
+    (z.map f).pmap g h' =
+    z.pmap (fun a ha => g (f a) (h' (f a) (mem_map.mpr ⟨a, ha, rfl⟩))) (fun a ha => ha) := by
+  induction' z with x y
+  rfl
+
+lemma pmap_map {P : α → Prop} {Q : β → Prop} (f : ∀ a, P a → β) (g : β → γ)
+    (z : Sym2 α) (h : ∀ a ∈ z, P a) (h' : ∀ b ∈ z.pmap f h, Q b) :
+    (z.pmap f h).map g = z.pmap (fun a ha => g (f a (h a ha))) (fun a ha ↦ ha) := by
+  induction' z with x y
+  rfl
+
+lemma pmap_pmap {P : α → Prop} {Q : β → Prop} (f : ∀ a, P a → β) (g : ∀ b, Q b → γ)
+    (z : Sym2 α) (h : ∀ a ∈ z, P a) (h' : ∀ b ∈ z.pmap f h, Q b) :
+    (z.pmap f h).pmap g h' = z.pmap (fun a ha => g (f a (h a ha))
+    (h' _ ((mem_pmap_iff f z h _).mpr ⟨a, ha, rfl⟩))) (fun a ha ↦ ha) := by
+  induction' z with x y
+  rfl
+
+/--
+"Attach" a proof `P a` that holds for all the elements of `s` to produce a new Sym2 object
+with the same elements but in the type `{x // P x}`.
+-/
+def attachWith {P : α → Prop} (s : Sym2 α)  (f : ∀ a ∈ s, P a) : Sym2 {a // P a} :=
```

Comment:
This would require a substantial reworking of this PR, but is it any easier to first define `attachWith` and then define `pmap s f` = `s.attach.map f`? This would enable you to rely on the `map` API to hopefully simplify the proofs about `pmap`.

Reply:
Would that be easier? Defining `attachWith` would be just a matter of adapting the current definition of pmap so that wouldn't be hard. But given that, at least the proofs I had to write so far, are mostly just `induction'` and `rfl`, I don't know this alternate approach can make things easier. Perhaps there is a value in using function on subtype rather than data and a proof as 2 inputs?